src/signatures/ternary/wots.rs

```diff
@@ -0,0 +1,367 @@
+
+        Ok(Self {
+            state,
+            marker: PhantomData,
+        })
+    }
+
+    fn as_trits(&self) -> &Trits<T1B1> {
+        &self.state
+    }
+}
+
+impl<S: Sponge + Default> WotsPrivateKey<S> {
+    /// Returns the inner trits.
+    pub fn as_trits(&self) -> &Trits<T1B1> {
+        &self.state
+    }
+}
+
+/// Winternitz One Time Signature public key.
+pub struct WotsPublicKey<S> {
+    state: TritBuf<T1B1Buf>,
+    marker: PhantomData<S>,
+}
+
+impl<S: Sponge + Default> PublicKey for WotsPublicKey<S> {
+    type Signature = WotsSignature<S>;
+    type Error = Error;
+
+    fn verify(&self, message: &Trits<T1B1>, signature: &Self::Signature) -> Result<bool, Self::Error> {
+        Ok(signature.recover_public_key(message)?.state == self.state)
+    }
+
+    fn size(&self) -> usize {
+        self.state.len()
+    }
+
+    fn from_trits(state: TritBuf<T1B1Buf>) -> Result<Self, Self::Error> {
+        if state.len() != HASH_LENGTH {
+            return Err(Error::InvalidPublicKeyLength(state.len()));
+        }
+
+        Ok(Self {
+            state,
+            marker: PhantomData,
+        })
+    }
+
+    fn as_trits(&self) -> &Trits<T1B1> {
+        &self.state
+    }
+}
+
+impl<S: Sponge + Default> Display for WotsPublicKey<S> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            self.as_trits().iter_trytes().map(char::from).collect::<String>()
+        )
+    }
+}
+
+/// Winternitz One Time Signature signature.
+pub struct WotsSignature<S> {
+    state: TritBuf<T1B1Buf>,
+    marker: PhantomData<S>,
+}
+
+impl<S: Sponge + Default> Signature for WotsSignature<S> {
+    type Error = Error;
+
+    fn size(&self) -> usize {
+        self.state.len()
+    }
+
+    fn from_trits(state: TritBuf<T1B1Buf>) -> Result<Self, Self::Error> {
+        if state.len() % SIGNATURE_FRAGMENT_LENGTH != 0 {
+            return Err(Error::InvalidSignatureLength(state.len()));
+        }
+
+        Ok(Self {
+            state,
+            marker: PhantomData,
+        })
+    }
+
+    fn as_trits(&self) -> &Trits<T1B1> {
+        &self.state
+    }
+}
+
+impl<S: Sponge + Default> RecoverableSignature for WotsSignature<S> {
+    type PublicKey = WotsPublicKey<S>;
+    type Error = Error;
+
+    fn recover_public_key(
+        &self,
+        message: &Trits<T1B1>,
+    ) -> Result<Self::PublicKey, <Self as RecoverableSignature>::Error> {
+        if message.len() != HASH_LENGTH {
+            return Err(Error::InvalidMessageLength(message.len()));
+        }
+
+        let mut sponge = S::default();
+        let mut public_key_state = TritBuf::<T1B1Buf>::zeros(HASH_LENGTH);
+        let security = self.state.len() / SIGNATURE_FRAGMENT_LENGTH;
+        let mut digests = TritBuf::<T1B1Buf>::zeros(security * HASH_LENGTH);
+        let mut hashed_signature = self.state.clone();
+
+        for (f, fragment) in hashed_signature.chunks_mut(SIGNATURE_FRAGMENT_LENGTH).enumerate() {
```

Comment:
This is an exact code duplication from `sign`

Reply:
Not really, the deepest loop differs

Reply:
Can definitely factorise it though

Reply:
Yeah, well it's probably not worth it